File: JavaScript/leetcode75/src/main/Problem25_AsteroidCollision.js
/**
 * @param {number[]} asteroids
 * @return {number[]}
 */
var asteroidCollision = function (asteroids) {
  let st = [];
  for (let i = 0; i < asteroids.length; i++) {
    if (st.length === 0 || asteroids[i] > 0) {
      st.push(asteroids[i]);
    } else {
      while (
        st.length > 0 &&
        st[st.length - 1] > 0 &&
        st[st.length - 1] < Math.abs(asteroids[i])
      ) {
        st.pop();
      }
      if (st[st.length - 1] < 0 || st.length === 0) {
        st.push(asteroids[i]);
      }
      if (st.length > 0 && st[st.length - 1] == Math.abs(asteroids[i])) {
        st.pop();
      }
    }
  }
  let res = new Array(st.length);
  for (let i = st.length - 1; i >= 0; i--) {
    res[i] = st.pop();
  }
  return res;
};

module.exports = asteroidCollision;

File: Java/leetcode75/src/main/java/org/phoenix/leetcode/challenges/Problem25_AsteroidCollision.java
package org.phoenix.leetcode.challenges;

import java.util.Stack;

public class Problem25_AsteroidCollision {
    public int[] asteroidCollision(int[] asteroids) {
        Stack<Integer> stack = new Stack<>();
        for (int size : asteroids) {
            if (stack.isEmpty() || size > 0) {
                stack.push(size);
            } else {
                while (!stack.isEmpty() && stack.peek() > 0 && stack.peek() < Math.abs(size)) {
                    stack.pop();
                }
                if (stack.peek() < 0 && !stack.isEmpty()) {
                    stack.push(size);
                }
                if (stack.peek() == Math.abs(size)) {
                    stack.pop();
                }
            }
        }
        int[] res = new int[stack.size()];
        for (int i = stack.size() - 1; i >= 0; i--) {
            res[i] = stack.pop();
        }
        return res;
    }
}
